### PR TITLE
Catalyst supports dynamic_one_shot

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
Catalyst does not currently support hardware-like execution for dynamic circuits.

**Description of the Change:**
Modify `parse_native_mid_circuit_measurements` to handle measurements with a broadcast dimension.
Add special handling of terminal measurements of MCMs. 

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-60756]

### CODE REVIEW

File: pennylane/transforms/dynamic_one_shot.py
Comment (dime10) [reply pennylane/transforms/dynamic_one_shot.py:230]: Especially specifying the expected contents and format of results would be helpful to follow along the rest of the code 😅

**Review by dime10 — COMMENTED** [review]
Looks good, although unfortunately between all the squeezing, stacking, list packing and unpacking, and without already knowing the expected data format in different places I found it quite hard to review those kinds of changes and will mostly have to trust you about them 😅

File: pennylane/transforms/dynamic_one_shot.py
Comment (dime10) [reply pennylane/transforms/dynamic_one_shot.py:215]: If I understand correctly, bypassing postselect should be conditional on the `postselect_mode` parameter right?

File: pennylane/measurements/probs.py
Comment (vincentmr) [reply pennylane/measurements/probs.py:97]: I don't remember what prompted this change, but this is not required anymore so I'll revert the change.

File: pennylane/transforms/dynamic_one_shot.py
Comment (vincentmr) [reply pennylane/transforms/dynamic_one_shot.py:215]: Yes, it will be dependent on `postselect_mode`.

File: pennylane/transforms/dynamic_one_shot.py
Comment (vincentmr) [inline pennylane/transforms/dynamic_one_shot.py:212]: This needs to be `if` because the block changes both `new_measurements` and `new_operations`.

File: pennylane/transforms/dynamic_one_shot.py
Comment (vincentmr) [reply pennylane/transforms/dynamic_one_shot.py:216]: Thanks for pointing this out, I'll update `new_op`.

File: pennylane/transforms/dynamic_one_shot.py
Comment (vincentmr) [reply pennylane/transforms/dynamic_one_shot.py:230]: It is obsolete indeed. I'll update it.

**Review by vincentmr — COMMENTED** [review]
> Looks good, although unfortunately between all the squeezing, stacking, list packing and unpacking, and without already knowing the expected data format in different places I found it quite hard to review those kinds of changes and will mostly have to trust you about them 😅

Right, sorry about that. I think once all MCM-related PRs go in, this will be a good candidate for refactoring.

File: pennylane/transforms/dynamic_one_shot.py
Comment (dime10) [reply pennylane/transforms/dynamic_one_shot.py:212]: > This needs to be if because the block changes both new_measurements and new_operations.

Oh my bad, but it's still confusing because the op can't be both `MidMeasureOp` and `MidCircuitMeasure` right? At least these two instructions should be exclusive (which is currently not enforced by the code structure, in fact I think with Mudit's inheritance PR both of these branches would be triggered):

```py
new_measurements.append(qml.sample(MeasurementValue([op], lambda res: res)))
new_measurements.append(qml.sample(op.out_classical_tracers[0]))
```

So while I'm not sure exactly how these three branches are supposed to be executing, I still think it would be better that the two statements I listed be guaranteed to be exclusive.

File: pennylane/transforms/dynamic_one_shot.py
Comment (mudit2812) [inline pennylane/transforms/dynamic_one_shot.py:293]:
```suggestion
            if active_qjit:
```

File: pennylane/transforms/dynamic_one_shot.py
Comment (mudit2812) [inline pennylane/transforms/dynamic_one_shot.py:310]:
```suggestion
def gather_mcm_qjit(circuit_measurement, measurement, is_valid):
```

File: pennylane/transforms/dynamic_one_shot.py
Comment (mudit2812) [inline pennylane/transforms/dynamic_one_shot.py:292]: I originally changed this to `qml.math.stack` because `qml.math.array` fails when trying to create an array using a list/tuple of other arrays. There is an unrelated issue with the tests currently, so it's hard to tell if this is actually causing any trouble or not. Just bringing it up here in case any interface tests for `dynamic_one_shot` start failing.

File: pennylane/transforms/dynamic_one_shot.py
Comment (mudit2812) [inline pennylane/transforms/dynamic_one_shot.py:317]: This docstring does not match the function arguments. Also, in my opinion, `circuit_measurement` should be `measurement`, and `measurement` should be `mcm_samples`.

File: pennylane/transforms/dynamic_one_shot.py
Comment (mudit2812) [inline pennylane/transforms/dynamic_one_shot.py:339]: Optional suggestion.

```suggestion
    sum_valid = qml.math.sum(is_valid)
    count_1 = qml.math.sum(meas * is_valid)
    if isinstance(circuit_measurement, CountsMP):
        return {0: sum_valid - count_1, 1: count_1}
    if isinstance(circuit_measurement, ProbabilityMP):
        counts = qml.math.array(
            [sum_valid - count_1, count_1], like=qml.math.get_deep_interface(is_valid)
        )
```

File: tests/conftest.py
Comment (mudit2812) [inline tests/conftest.py:30]: I'm not sure why, but I think this may be the culprit for the failing tests.

**Review by mudit2812 — CHANGES_REQUESTED** [review]
Looks great so far! The failing CI seems problematic, and I left a few comments about the changes to the transform as well.

File: pennylane/transforms/dynamic_one_shot.py
Comment (vincentmr) [reply pennylane/transforms/dynamic_one_shot.py:292]: The stacking should already have happened in `processing_fn` now, so hopefully this problem doesn't reappear.

File: pennylane/transforms/dynamic_one_shot.py
Comment (vincentmr) [reply pennylane/transforms/dynamic_one_shot.py]: > You can also merge the code inside the elif with the very first condition in the loop, though I must admit that would make it harder to understand.

The latter was @dime10 's point for changing it to the current version.

File: tests/conftest.py
Comment (vincentmr) [reply tests/conftest.py:30]: Yes, I changed the module name.

File: pennylane/transforms/dynamic_one_shot.py
Comment (mudit2812) [reply pennylane/transforms/dynamic_one_shot.py:292]: Looks like the tests are still passing. Nice!

File: pennylane/transforms/dynamic_one_shot.py
Comment (mudit2812) [inline pennylane/transforms/dynamic_one_shot.py:334]: You can put these inside the `if isinstance(measurement, (CountsMP, ProbabilityMP)):` block.

**Review by mudit2812 — COMMENTED** [review]
Looks great, just one stylistic optional suggestion. Other than that, happy to approve once code coverage is passing.

**Comment by vincentmr** [comment]
> Other than that, happy to approve once code coverage is passing.

Code coverage will not pass until the CI can pull a Catalyst version supporting the one-shot mode. So should we force merge and come back to it once the Catalyst support is out?

**Comment by mudit2812** [comment]
> > Other than that, happy to approve once code coverage is passing.
> 
> Code coverage will not pass until the CI can pull a Catalyst version supporting the one-shot mode. So should we force merge and come back to it once the Catalyst support is out?

I'd prefer to avoid force merging.

You can add `# pragma: no cover` to the uncovered branches along with a `TODO` to remove them after the catalyst PR is merged. Probably worth creating a story for the follow up as well and adding it to the P0 epic.

**Review by mudit2812 — COMMENTED** [review]
Left a comment about using `# pragma: no cover` to silence the coverage errors. An alternative would be to mock `qml.compiler.active` to always return `True` and use the mocked function in unit tests.

All other changes look good to me and I'm happy to approve once this coversation is resolved.

**Comment by vincentmr** [comment]
Thanks @mudit2812 . I'll go with `# pragma: no cover` for now since the tests are already written and ready to cover once the Catalyst support comes in.

File: pennylane/transforms/dynamic_one_shot.py
Comment (mudit2812) [inline pennylane/transforms/dynamic_one_shot.py:53]: Python 3.9 added support for using standard collections as type hints so we don't need to use `typing.Tuple`. `tuple` should be fine, and I think it might even be better to start standardizing its use in our type hints.

```suggestion
) -> tuple[Sequence[qml.tape.QuantumTape], Callable]:
```

File: pennylane/transforms/dynamic_one_shot.py
Comment (mudit2812) [inline pennylane/transforms/dynamic_one_shot.py:22]: To follow up on the above comment.

```suggestion
from typing import Callable, Sequence
```

**Review by mudit2812 — APPROVED** [review]

File: pennylane/transforms/dynamic_one_shot.py
Comment (dime10) [inline pennylane/transforms/dynamic_one_shot.py:214]: I think I mentioned in some other pr, but in this case `new_op` is just a different name for the same object. We should either copy the op or remove the `new_op` reference.

File: pennylane/transforms/dynamic_one_shot.py
Comment (dime10) [inline pennylane/transforms/dynamic_one_shot.py:213]: I mentioned this on slack, but I don't think we should set this attribute in PennyLane to effect a behaviour in Catalyst, because it is rather random. For now, leaving it out entirely means we just always consider it False (i.e. using "fill-shots" mode), and when Mudit implements the config handling on the catalyst side the bypass can be effected there based on the config.

File: pennylane/transforms/dynamic_one_shot.py
Comment (dime10) [inline pennylane/transforms/dynamic_one_shot.py:301]: It would be helpful to explain such special handling with a comment :)

File: pennylane/transforms/dynamic_one_shot.py
Comment (dime10) [inline pennylane/transforms/dynamic_one_shot.py:335]: I think this breaks the return type convention of counts in catalyst, which is not a dictionary. I'm not sure how serious this is though, but if in doubt would this be acceptable? 

```suggestion
            return qml.math.array([0, 1]), qml.math.array([sum_valid - count_1, count_1])
```

Or does it break further processing code?

**Review by dime10 — CHANGES_REQUESTED** [review]
Looks great, just a couple of minor things:

File: pennylane/transforms/dynamic_one_shot.py
Comment (vincentmr) [reply pennylane/transforms/dynamic_one_shot.py:213]: Now that the "fill-shots" mode exists, I'm all for it.

File: pennylane/transforms/dynamic_one_shot.py
Comment (vincentmr) [reply pennylane/transforms/dynamic_one_shot.py:214]: Will remove following the introduction of `postselection_mode`.

File: pennylane/transforms/dynamic_one_shot.py
Comment (vincentmr) [reply pennylane/transforms/dynamic_one_shot.py:297]: Will do.

File: pennylane/transforms/dynamic_one_shot.py
Comment (vincentmr) [reply pennylane/transforms/dynamic_one_shot.py]: This was to sync with PennyLane as much as possible, but I can return arrays without issue. I'll just have to fix how the tests validate measurements.

**Review by dime10 — APPROVED** [review]
Happy to approve 💯

File: pennylane/transforms/dynamic_one_shot.py
Comment (mudit2812) [inline pennylane/transforms/dynamic_one_shot.py:211]: Just had a realization. We could store the measurement value/tracer in `MidMeasureMP.mv` once `MidCircuitMeasure` inherits `MidMeasureMP` and then we can get rid of this branch. Not needed right now, but I can come back and update this once the Catalyst side work is complete.

File: pennylane/transforms/dynamic_one_shot.py
Comment (mudit2812) [inline pennylane/transforms/dynamic_one_shot.py:213]: Just had a realization. We could store the measurement value/tracer in `MidMeasureMP.mv` once `MidCircuitMeasure` inherits `MidMeasureMP` and then we can get rid of this branch. Not needed right now, but I can come back and update this once the Catalyst side work is complete. Then this line could instead be the following, and we won't need another branch for `MidMeasureMP`.
```suggestion
        if isinstance(op, MidMeasureMP):
            new_measurements.append(qml.sample(op=op.mv))
```

```suggestion
        if "MidCircuitMeasure" in str(type(op)):  # pragma: no cover
            new_measurements.append(qml.sample(op.out_classical_tracers[0]))
        elif isinstance(op, MidMeasureMP):
            new_measurements.append(qml.sample(MeasurementValue([op], lambda res: res)))
```

Comment (mudit2812) [reply pennylane/transforms/dynamic_one_shot.py:213]: This change will not work right now though.

**Pull request merged**
